### PR TITLE
Add protobuf api integrated with last fetched weather

### PR DESCRIPTION
**Pull request by Vanshbajaj**
Added Protocol Buffer Api 
Prevoius Fetch Weather Shown

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -4,6 +4,5 @@ plugins {
     alias(libs.plugins.kotlin.android) apply false
     alias(libs.plugins.android.library) apply false
     alias(libs.plugins.hilt.android)  apply false
-
-
-}
\ No newline at end of file
+    id("com.google.protobuf") version "0.9.1" apply  false
```
Comment (shubhamsinghmutualmobile) [inline build.gradle.kts:7]: Convert it to a version catalog entry (to keep all dependencies looking consistent)

File: data/build.gradle.kts [hunk 2]
```diff
@@ -6,6 +7,7 @@ plugins {
     alias(libs.plugins.kotlin.android)
     id("kotlin-kapt")
     id("dagger.hilt.android.plugin")
+    id("com.google.protobuf")
```
Comment (shubhamsinghmutualmobile) [inline data/build.gradle.kts:10]: Same for these too

File: data/build.gradle.kts [hunk 3]
```diff
@@ -64,5 +81,31 @@ dependencies {
...
+    implementation(libs.protobuf.javalite)
 
 }
+kapt {
+    correctErrorTypes= true
+}
+
+protobuf {
+    protoc {
+        artifact = "com.google.protobuf:protoc:3.21.0" // Adjust the version
```
Comment (shubhamsinghmutualmobile) [inline data/build.gradle.kts:95]: Same

File: data/build.gradle.kts [hunk 4]
```diff
@@ -64,5 +81,31 @@ dependencies {
...
+    }
+    generateProtoTasks {
+        all().forEach { task ->
+            task.builtins {
+                create("java") {
+                    option("lite")
+                }
+//                create("kotlin") {
+//                    option("lite")
+//                }
```
Comment (shubhamsinghmutualmobile) [inline data/build.gradle.kts:105]: Remove commented/dead code

File: data/src/main/java/com/open/data/proto/DataStoreModule.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import android.content.Context
+import androidx.datastore.core.DataStore
+import androidx.datastore.dataStore
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.android.qualifiers.ApplicationContext
+import dagger.hilt.components.SingletonComponent
+private val Context.dataStore: DataStore<MyData> by dataStore("my_data_store", MyDataSerializer)
+@Module
```
Comment (shubhamsinghmutualmobile) [inline data/src/main/java/com/open/data/proto/DataStoreModule.kt:12]: Improve spacing b/w different sections of the code

File: data/src/main/java/com/open/data/proto/MyDataSerializer.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    override suspend fun readFrom(input: InputStream): MyData {
+        return MyData.parseFrom(input)
+    }
+
+    override suspend fun writeTo(t: MyData, output: OutputStream) {
+        t.writeTo(output)
+    }
+}
+
+
```
Comment (shubhamsinghmutualmobile) [inline data/src/main/java/com/open/data/proto/MyDataSerializer.kt:22]: Remove dual spacing between code sections (keep spacing consistent)

File: presentation/build.gradle.kts [hunk 7]
```diff
@@ -41,12 +41,14 @@ android {
...
 dependencies {
-
     implementation(libs.androidx.core.ktx)
     implementation(libs.androidx.appcompat)
     implementation(libs.material)
     implementation(libs.androidx.runtime.android)
     implementation(libs.androidx.ui.text.android)
+    implementation(libs.androidx.datastore)
+    implementation(libs.protobuf.javalite)
+    implementation("com.google.protobuf:protobuf-kotlin-lite:3.21.5")
```
Comment (shubhamsinghmutualmobile) [inline presentation/build.gradle.kts:51]: Same, convert this to a version catalog entry

File: presentation/src/main/java/com/open/presentation/ui/components/Extention.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.open.presentation.ui.components
+
+import androidx.datastore.core.DataStore
+import androidx.datastore.dataStore
+import com.open.data.proto.MyData
+import com.open.data.proto.MyDataSerializer
+
+class Extention {
+    private val android.content.Context.dataStore: DataStore<MyData> by dataStore("my_data_store", MyDataSerializer)
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/open/presentation/ui/components/Extention.kt:9]: It is a `data` module component, and does not belong in `presentation`

File: presentation/src/main/java/com/open/presentation/ui/screens/WeatherScreen.kt [hunk 9]
```diff
@@ -78,7 +84,17 @@ fun WeatherScreen(
             }
 
             is UiState.Empty -> {
-                Text(text = "Enter a city to get weather information")
+                LaunchedEffect(Unit) {
+                    viewModel.viewModelScope.launch {
+
+                    }
+
+                }
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/open/presentation/ui/screens/WeatherScreen.kt:92]: Remove dead code

File: presentation/src/main/java/com/open/presentation/ui/screens/WeatherScreen.kt [hunk 10]
```diff
@@ -78,7 +84,17 @@ fun WeatherScreen(
...
+                val name = viewModel.nameFlow.collectAsState("").value
+                val temp = viewModel.temp.collectAsState("").value
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/open/presentation/ui/screens/WeatherScreen.kt:94]: Do these states need to be inside the viewModel? Can we not have them be local rememberSaveable states?

File: presentation/src/main/java/com/open/presentation/ui/viewmodel/WeatherViewModel.kt [hunk 11]
```diff
@@ -1,25 +1,47 @@
...
 import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flow
 import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class WeatherViewModel @Inject constructor(
-    private val getWeatherUseCase: GetWeatherUseCase
+    private val getWeatherUseCase: GetWeatherUseCase, private val dataStore: DataStore<MyData>
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/open/presentation/ui/viewmodel/WeatherViewModel.kt:26]: Nit: Put these on different lines

File: presentation/src/main/java/com/open/presentation/ui/viewmodel/WeatherViewModel.kt [hunk 12]
```diff
@@ -1,25 +1,47 @@
...
 
+
     private val _weatherState = MutableStateFlow<UiState<Weather>>(UiState.Empty)
     val weatherState: StateFlow<UiState<Weather>> get() = _weatherState
+    private val _nameFlow = MutableStateFlow("")
+    private val _temp = MutableStateFlow(0)
+
+
+
+    init {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/open/presentation/ui/viewmodel/WeatherViewModel.kt:37]: Improve code spacing to be more consistent/standard

**Review by shubhamsinghmutualmobile — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/open/presentation/ui/screens/WeatherScreen.kt (hunk 10 above)
Comment (Vanshbajaj) [reply presentation/src/main/java/com/open/presentation/ui/screens/WeatherScreen.kt:94]: as these states are fetch  from ViewModel that is why using from viewmodel

**Review by shubhamsinghmutualmobile — APPROVED** [review]

**Pull request merged**